runtime/major_gc.c

```diff
@@ -148,10 +148,61 @@ static void realloc_gray_vals (void)
   }
 }
 
+#ifdef NATIVE_CODE_AND_NO_NAKED_POINTERS
+
+static int is_pointer_safe (value v, value *p)
+{
+  volatile header_t h;
+  tag_t t;
+
+  Caml_state->checking_pointer_pc = &&on_segfault;
+  h = Hd_val(v);
+  Caml_state->checking_pointer_pc = NULL;
+
+  t = Tag_val(v);
+  if (t == Infix_tag){
+    v -= Infix_offset_val(v);
+    h = Hd_val (v);
+    t = Tag_hd(h);
+  }
+
+  /* For the pointer to be considered safe, either the given pointer is in heap
+   * or the (out of heap) pointer has a black header and its size is < 2 ** 40
+   * words (128 GB). If not, we report a warning. */
+  if (Is_in_heap (v) || (Is_black_hd(h) && Wosize_hd(h) < (1ul << 40)))
+    return 1;
+
+on_segfault:
+  if (Caml_state->checking_pointer_pc == NULL) {
+    if (!Is_black_hd(h)) {
+      fprintf (stderr, "Out-of-heap pointer at %p of value %p has "
+                      "non-black head (tag=%d)\n", p, (void*)v, t);
+    } else {
+      fprintf (stderr, "Out-of-heap pointer at %p of value %p has "
+                       "suspiciously large size: %lu words\n",
+               p, (void*)v, Wosize_hd(h));
+    }
+  } else {
+    Caml_state->checking_pointer_pc = NULL;
+    fprintf (stderr, "Out-of-heap pointer at %p of value %p. "
+                     "Cannot read head.\n", p, (void*)v);
```

Comment:
I was naively expecting `on_segfault` to jump right here.

Reply:
Same as above.